lib/puppet/provider/service/smf.rb

```diff
@@ -2,117 +2,139 @@
 
 # Solaris 10 SMF-style services.
 Puppet::Type.type(:service).provide :smf, :parent => :base do
+  # TODO: Update the documentation with things that you find.
+  # E.g. About specifying fmri explicitly in the manifest.
   desc <<-EOT
     Support for Sun's new Service Management Framework.
 
-    Starting a service is effectively equivalent to enabling it, so there is
-    only support for starting and stopping services, which also enables and
-    disables them, respectively.
-
     By specifying `manifest => "/path/to/service.xml"`, the SMF manifest will
     be imported if it does not exist.
-
   EOT
 
   defaultfor :osfamily => :solaris
 
   confine :osfamily => :solaris
 
-  commands :adm => "/usr/sbin/svcadm", :svcs => "/usr/bin/svcs"
-  commands :svccfg => "/usr/sbin/svccfg"
+  commands :adm => "/usr/sbin/svcadm",
+           :svcs => "/usr/bin/svcs",
+           :svccfg => "/usr/sbin/svccfg"
 
   has_feature :refreshable
 
-  def setupservice
-      if resource[:manifest]
-        begin
-          svcs("-l", @resource[:name])
-        rescue Puppet::ExecutionFailure
-          Puppet.notice "Importing #{@resource[:manifest]} for #{@resource[:name]}"
-          svccfg :import, resource[:manifest]
-        end
-      end
+  def self.instances
+   service_instances = svcs("-H", "-o", "state,fmri" ).split("\n")
+
+   # Puppet does not manage services in the legacy_run state, so filter those out.
+   service_instances.reject! { |line| line =~ /^legacy_run/ }
+   
+   service_instances.collect! do |line|
+     _state, fmri = line.split(/\s+/)
+     new({:name => fmri})
```

Comment:
This should be fine, some of the other providers return only the resource name so I thought it'd be good to be consistent with them (see e.g. https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/service/systemd.rb#L37 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/service/windows.rb#L104). The previously returned FMRI wasn't used anywhere either from what I could find. Plus, I only see this code being used here: https://github.com/ekinanp/puppet/blob/master/lib/puppet/indirector/resource/ral.rb#L12-L18 and that code queries only for the name (but Puppet's a big codebase, so I'm more than likely wrong there).